packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts

```diff
@@ -0,0 +1,225 @@
+        );
+
+      expect(toJSON(item)).containDeep(
+        toJSON({
+          id: item.id,
+          description: 'an item',
+        }),
+      );
+
+      const persistedItem = await cartItemRepo.findById(item.id);
+      expect(toJSON(persistedItem)).to.deepEqual(toJSON(item));
+      const persistedLink = await customerCartItemLinkRepo.find();
+      expect(toJSON(persistedLink[0])).to.containDeep(
+        toJSON({
+          customerId: existingCustomerId,
+          cartItemId: item.id,
+          description: 'a through model',
+        }),
+      );
+    });
+
+    it('can find instances of the related model', async () => {
+      const item = await customerRepo
+        .cartItems(existingCustomerId)
+        .create(
+          {description: 'an item'},
+          {throughData: {description: 'a through model'}},
+        );
+      const notMyItem = await cartItemRepo.create({
+        description: "someone else's item",
+      });
+      const result = await customerRepo.cartItems(existingCustomerId).find();
+      expect(toJSON(result)).to.containDeep(toJSON([item]));
+      expect(toJSON(result[0])).to.not.containEql(toJSON(notMyItem));
+    });
+
+    it('can patch an instance', async () => {
+      const item1 = await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+      const item2 = await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+      const count = await customerRepo
+        .cartItems(existingCustomerId)
+        .patch({description: 'group 2'}, {id: item2.id});
+
+      expect(count.count).to.equal(1);
+      const result = await customerRepo.cartItems(existingCustomerId).find();
+      expect(toJSON(result)).to.containDeep(
+        toJSON([
+          {id: item1.id, description: 'group 1'},
+          {id: item2.id, description: 'group 2'},
+        ]),
+      );
+    });
+
+    it('throws error when query tries to change the target id', async () => {
+      // a diff id for CartItem instance
+      const anotherId = (await givenPersistedCustomerInstance()).id;
+      await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+
+      await expect(
+        customerRepo.cartItems(existingCustomerId).patch({id: anotherId}),
+      ).to.be.rejectedWith(/Property "id" cannot be changed!/);
+    });
+
+    it('can delete many instances and their through models', async () => {
+      await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+      await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+
+      let links = await customerCartItemLinkRepo.find();
+      let cartItems = await cartItemRepo.find();
+      expect(links).have.length(2);
+      expect(cartItems).have.length(2);
+
+      await customerRepo.cartItems(existingCustomerId).delete();
```

Comment:
nitpick: does `delete` takes in a filter? Maybe add another test for delete with filter

Reply:
Just realized that the `delete` method has a bug. It is able to delete **targets** based on the filter but it deletes **all through instance that has the souceKey**. Will open up an issue to fix it.

Reply:
The issue is being tracked in https://github.com/strongloop/loopback-next/issues/5852